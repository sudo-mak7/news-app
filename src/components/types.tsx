import { CommentsInterface } from '@common/types/commentsInterface'

export interface AnswerSectionComponentInterface {
  isCollapsed: boolean,
  answers: CommentsInterface[],
  isLoading: boolean
}