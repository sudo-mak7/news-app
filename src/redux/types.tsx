import * as React from 'react'
import { CommentsInterface } from '@common/types/commentsInterface'
import { NewsInterface } from '@common/types/newsInterface'

export interface fetchAnswersInterface {
  id: number,
  setAnswersState: React.Dispatch<React.SetStateAction<CommentsInterface[]>>,
  setIsLoading: React.Dispatch<React.SetStateAction<boolean>>,
  setError: React.Dispatch<React.SetStateAction<any>>
}

export interface CommentsStateInterface {
  comments: CommentsInterface[],
  loading: boolean,
  error: any
}

export interface CurrentNewsStateInterface {
  currentNews: NewsInterface,
  loading: boolean,
  error: any
}

export interface NewsByIdStateInterface {
  id: number
}

export interface NewsIdsStateInterface {
  newsIds: [number][],
  loading: boolean,
  error: any
}

export interface NewsStateInterface {
  news: NewsInterface[],
  loading: boolean,
  error: any
}

export interface PaginationStateInterface {
  currentPage: number,
  currentPageNews: NewsInterface[],
  pagesLeft: number
}