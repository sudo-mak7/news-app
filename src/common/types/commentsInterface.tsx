interface Comments {
  by: string,
  id: number,
  parent: number,
  text: string,
  time: number,
  type: string
  kids: number[],
  dead: boolean,
  deleted: boolean
}

export type CommentsInterface = Partial<Comments>