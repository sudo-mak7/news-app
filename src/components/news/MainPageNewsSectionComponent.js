import React from 'react'
import {
  Header,
  Icon,
  Label,
  Segment
} from 'semantic-ui-react'
import { Link } from 'react-router-dom'
import { useDispatch } from 'react-redux'
import { setCurrentNewsId } from '@redux/news/newsByIdSlice'
import { dateNormalizer } from '@utils/dateNormalizer'

const MainPageNewsSectionComponent = ({ title, by, score, time, id }) => {
  const dispatch = useDispatch()

  const date = dateNormalizer(time)

  const getNewsId = (e) => {
    localStorage.removeItem('currentNewsId')
    const id = e.currentTarget.id
    dispatch(setCurrentNewsId(id))
    localStorage.setItem('currentNewsId', JSON.stringify(id))
  }

  return (
    <Segment.Group
      id={ id }
      onClick={ (e) => getNewsId(e) }
    >
      <Link to='/news-page'>
        <Segment attached='top' color='blue'>
          <Header as='h2' color='black'>
            { title || 'No title' }
          </Header>
        </Segment>

        <Segment attached='bottom' clearing>
          <Label image floated='left'>
            <Icon name='user'></Icon>
            { by || 'No author' }
          </Label>
          <Label>
            <Icon name='star'></Icon>
              Rating
            <Label.Detail>
              { score || 'No rating' }
            </Label.Detail>
          </Label>
          <Label>
            <Icon name='calendar'></Icon>
            { date || 'No date' }
          </Label>
        </Segment>
      </Link>
    </Segment.Group>
  )
}

export default MainPageNewsSectionComponent