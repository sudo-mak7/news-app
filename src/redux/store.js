import { legacy_createStore as createStore, compose, applyMiddleware } from 'redux'
import createSagaMiddleware from 'redux-saga'
import rootReducer from './reducers/rootReducer'
import { newsWatcher } from './sagas/rootSaga'

const sagaMiddleware = createSagaMiddleware()

const composeEnhancers =
  typeof window === 'object' && window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__
    ? window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__({})
    : compose

const configureStore = (preloadedState) => createStore(
  rootReducer,
  preloadedState,
  composeEnhancers(applyMiddleware(sagaMiddleware))
)

const store = configureStore({})

sagaMiddleware.run(newsWatcher)

export default store
