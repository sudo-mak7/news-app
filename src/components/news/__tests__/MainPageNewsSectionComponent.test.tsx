import * as React from 'react'
import * as reduxHooks from '@redux/reduxHooks'
import { render, screen, fireEvent } from '@testing-library/react'
import { BrowserRouter } from 'react-router-dom'
import { Provider } from 'react-redux'
import { store } from '@redux/store'
import MainPageNewsSectionComponent from '@components/news/MainPageNewsSectionComponent'
import { NewsInterface } from '@common/types/newsInterface'

const mainPageNewsSectionComponent = (
  { title, by, score, time, id }: NewsInterface
): JSX.Element =>
  <Provider store = { store } >
    <BrowserRouter>
      <MainPageNewsSectionComponent
        title={ title }
        by={ by }
        score={ score }
        time={ time }
        id={ id }
      />
    </BrowserRouter>
  </Provider>

const mockedDispatch = jest.spyOn(reduxHooks, 'useAppDispatch')

describe('MainPageNewsSectionComponent test', () => {
  test('MainPageNewsSectionComponent should to match snapshot', () => {
    const props = {
      title: 'News title',
      by: 'author',
      score: 111,
      time: 1175714200,
      id: 8863
    }

    render(mainPageNewsSectionComponent(props))
    expect(mainPageNewsSectionComponent(props)).toMatchSnapshot()
  })

  test('renders MainPageNewsSectionComponent with data', () => {
    const props = {
      title: 'News title',
      by: 'author',
      score: 111,
      time: 1175714200,
      id: 8863
    }

    render(mainPageNewsSectionComponent(props))

    const newsInMainPageNewsSectionComponent = screen.getByTestId('news')
    expect(newsInMainPageNewsSectionComponent).toBeInTheDocument()
  })

  test('renders MainPageNewsSectionComponent with no data', () => {
    const props = {
      title: null,
      by: null,
      score: null,
      time: null,
      id: null
    }

    render(mainPageNewsSectionComponent(props))

    const titleInMainPageNewsSectionComponent = screen.getByText(/no title/i)
    expect(titleInMainPageNewsSectionComponent).toBeInTheDocument()

    const authorInMainPageNewsSectionComponent = screen.getByText(/no author/i)
    expect(authorInMainPageNewsSectionComponent).toBeInTheDocument()

    const ratingInMainPageNewsSectionComponent = screen.getByText(/no rating/i)
    expect(ratingInMainPageNewsSectionComponent).toBeInTheDocument()

    const dateInMainPageNewsSectionComponent = screen.getByText(/no date/i)
    expect(dateInMainPageNewsSectionComponent).toBeInTheDocument()
  })

  test('dispatch should called once by click', () => {
    const props = {
      title: 'News title',
      by: 'author',
      score: 111,
      time: 1175714200,
      id: 8863
    }

    const dispatch = jest.fn()
    mockedDispatch.mockReturnValue(dispatch)

    render(mainPageNewsSectionComponent(props))
    const newsInMainPageNewsSectionComponent = screen.getByTestId('news')

    fireEvent.click(newsInMainPageNewsSectionComponent)
    expect(dispatch).toHaveBeenCalledTimes(1)
  })

  afterAll(() => {
    jest.clearAllMocks()
  })
})