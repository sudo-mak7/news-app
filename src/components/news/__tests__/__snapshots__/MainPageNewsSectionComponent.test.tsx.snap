// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MainPageNewsSectionComponent test MainPageNewsSectionComponent should to match snapshot 1`] = `
<Provider
  store={
    Object {
      "@@observable": [Function],
      "dispatch": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <BrowserRouter>
    <MainPageNewsSectionComponent
      by="author"
      id={8863}
      score={111}
      time={1175714200}
      title="News title"
    />
  </BrowserRouter>
</Provider>
`;

exports[`MainPageNewsSectionComponent test renders MainPageNewsSectionComponent with data 1`] = `
<Provider
  store={
    Object {
      "@@observable": [Function],
      "dispatch": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <BrowserRouter>
    <MainPageNewsSectionComponent
      by="author"
      id={8863}
      score={111}
      time={1175714200}
      title="News title"
    />
  </BrowserRouter>
</Provider>
`;

exports[`MainPageNewsSectionComponent test renders MainPageNewsSectionComponent with no data 1`] = `
<Provider
  store={
    Object {
      "@@observable": [Function],
      "dispatch": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <BrowserRouter>
    <MainPageNewsSectionComponent
      by={null}
      id={null}
      score={null}
      time={null}
      title={null}
    />
  </BrowserRouter>
</Provider>
`;
